nixos/modules/services/databases/neo4j.nix

```diff
@@ -36,81 +89,414 @@ let
+
+    # Usage Data Collector
+    dbms.udc.enabled=${boolToString cfg.udc.enable}
+
+    # Extra Configuration
+    ${cfg.extraServerConfig}
   '';
 
 in {
 
   ###### interface
 
   options.services.neo4j = {
-    enable = mkOption {
-      description = "Whether to enable neo4j.";
+
+    allowUpgrade = mkOption {
+      type = types.bool;
       default = false;
+      description = ''
+        Allow upgrade of Neo4j database files from an older version.
+      '';
+    };
+
+    constrainLoadCsv = mkOption {
       type = types.bool;
+      default = true;
+      description = ''
+        Sets the root directory for file URLs used with the Cypher
+        <literal>LOAD CSV</literal> clause to be that defined by
+        <literal>services.neo4j.directories.imports</literal>. It restricts
+        access to only those files within that directory and its subdirectories.
+
+        Setting this option to <literal>false</literal> introduces possible
+        security problems.
+      '';
+    };
+
+    defaultListenAddress = mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      description = ''
+        Default network interface to listen for incoming connections. To
+        listen for connections on all interfaces, use "0.0.0.0".
+
+        Specifies the default IP address and address part of connector specific
+        <literal>listenAddress</literal> options. To bind specific connectors
+        to a specific network interfaces, specify the entire
+        <literal>listenAddress</literal> option for that connector.
+      '';
+    };
+
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to enable Neo4j Community Edition.
+      '';
+    };
+
+    extraServerConfig = mkOption {
+      type = types.lines;
+      default = "";
+      description = ''
+        Extra configuration for Neo4j Community server. Refer to the
+        <link xlink:href="https://neo4j.com/docs/operations-manual/current/reference/configuration-settings/">complete reference</link>
+        of Neo4j configuration settings.
+      '';
     };
 
     package = mkOption {
-      description = "Neo4j package to use.";
+      type = types.package;
       default = pkgs.neo4j;
       defaultText = "pkgs.neo4j";
-      type = types.package;
+      description = ''
+        Neo4j package to use.
+      '';
     };
 
-    listenAddress = mkOption {
-      description = "Neo4j listen address.";
-      default = "127.0.0.1";
-      type = types.str;
+    readOnly = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Only allow read operations from this Neo4j instance.
+      '';
     };
 
-    port = mkOption {
-      description = "Neo4j port to listen for HTTP traffic.";
-      default = 7474;
-      type = types.int;
+    workerCount = mkOption {
+      type = types.either (types.enum [ "Available Processors" ]) (types.ints.between 1 44738);
```

Comment:
I think it would be better to just use `types.nullOr`, `null` signifying all available processors. Also why is this number limited to exactly 44738?

Reply:
This is the valid range of values for the option as described by the [configuration setting](https://neo4j.com/docs/operations-manual/3.4/reference/configuration-settings/#config_dbms.threads.worker_count). I chose an enum to make it easy to understand what the actual Neo4j config setting defaults to. Using `nullOr` with mention of what the `null` means in the docstring works too, I just thought using the enum would be a more direct way of saying so.